### PR TITLE
Update controller-runtime to v0.16.4

### PR DESCRIPTION
**Pull request by bshephar**
This change updates controller-runtime and associated dependencies

### CODE REVIEW

File: modules/ansible/go.mod [hunk 1]
```diff
@@ -1,6 +1,6 @@
 module github.com/openstack-k8s-operators/lib-common/modules/ansible
 
-go 1.19
+go 1.21
```
Comment (gibizer) [inline modules/ansible/go.mod:3]: add toolchain

**Comment by bshephar** [comment]
Hmm, looks like the CI env isn't using Go 1.21:
https://github.com/openstack-k8s-operators/lib-common/actions/runs/7762273514/job/21172396751?pr=413#step:5:819

```
ERRO Running error: context loading failed: failed to load packages: failed to load with go/packages: err: exit status 1: stderr: go: errors parsing go.mod:
/home/runner/work/lib-common/lib-common/modules/common/go.mod:5: unknown directive: toolchain 
```

**Comment by lewisdenny** [comment]
This is just using Github actions, not Prow or Zuul, the Github workflow will need to be updated AFAIU

**Comment by lewisdenny** [comment]
Updating that line in this PR should trigger the Github action to run with 1.21

**Comment by lewisdenny** [comment]
Added the Golang version override to the pre-commit job as well

**Comment by lewisdenny** [comment]
Nice that helped! 💚

**Review by gibizer — APPROVED** [review]

**Review by abays — APPROVED** [review]
/lgtm

**Pull request merged**